File: Union Find Algorithm/C/union_find.c
#include <stdio.h>
#include <stdlib.h>

// Structure to represent a subset
struct Subset {
    int parent;
    int rank;
};

// Function to initialize subsets
void makeSet(struct Subset subsets[], int n) {
    for (int i = 0; i < n; i++) {
        subsets[i].parent = i;
        subsets[i].rank = 0;
    }
}

// Find the root of a given element 'x'
int find(struct Subset subsets[], int x) {
    if (subsets[x].parent != x)
        subsets[x].parent = find(subsets, subsets[x].parent); // Path compression
    return subsets[x].parent;
}

// Union two subsets based on rank
void unionSets(struct Subset subsets[], int x, int y) {
    int rootX = find(subsets, x);
    int rootY = find(subsets, y);

    if (rootX != rootY) {
        if (subsets[rootX].rank < subsets[rootY].rank)
            subsets[rootX].parent = rootY;
        else if (subsets[rootX].rank > subsets[rootY].rank)
            subsets[rootY].parent = rootX;
        else {
            subsets[rootY].parent = rootX;
            subsets[rootX].rank++;
        }
    }
}

int main() {
    int n = 6; // Number of elements

    struct Subset subsets[n];
    makeSet(subsets, n);

    // Perform some union operations
    unionSets(subsets, 0, 1);
    unionSets(subsets, 1, 2);
    unionSets(subsets, 3, 4);

    // Check if elements are in the same subset
    printf("Are 0 and 2 in the same subset? %s\n", (find(subsets, 0) == find(subsets, 2)) ? "Yes" : "No");
    printf("Are 0 and 4 in the same subset? %s\n", (find(subsets, 0) == find(subsets, 4)) ? "Yes" : "No");

    return 0;
}

File: website/algorithms/union-find-algorithm/union_find.py
class UnionFind:
    def __init__(self, size):
        self.parent = list(range(size))
        self.rank = [1] * size

    def find(self, p):
        if self.parent[p] != p:
            self.parent[p] = self.find(self.parent[p])  # Path compression
        return self.parent[p]

    def union(self, p, q):
        rootP = self.find(p)
        rootQ = self.find(q)
        if rootP != rootQ:
            if self.rank[rootP] > self.rank[rootQ]:
                self.parent[rootQ] = rootP
            elif self.rank[rootP] < self.rank[rootQ]:
                self.parent[rootP] = rootQ
            else:
                self.parent[rootQ] = rootP
                self.rank[rootP] += 1

# Example usage
uf = UnionFind(10)
uf.union(1, 2)
uf.union(3, 4)
print(uf.find(1))  # Output: 1
print(uf.find(2))  # Output: 1
print(uf.find(3))  # Output: 3
print(uf.find(4))  # Output: 3
